dmoj/judge.py

```diff
@@ -459,69 +491,92 @@ def _grade_cases(self) -> Generator[Tuple[IPC, tuple], None, None]:
+                    global_test_number += 1
+                    yield IPC.RESULT, (None, global_test_number, grade_one(test))
 
-                    # If the submission was killed due to a user-initiated abort, any result is meaningless.
-                    if self._abort_requested:
-                        yield IPC.GRADING_ABORTED, ()
-                        return
+                if skip_current:
+                    failed_tests.add(test.config.raw_config_id)
+                    if not test.points or self.submission.short_circuit:
+                        skip_all = True
+
+            return failed_tests
+
+        def grade_one(case) -> Result:
+            nonlocal skip_current
+            if skip_current:
+                return Result(case, result_flag=Result.SC)
+            else:
+                result = self.grader.grade(case)
 
-                    if result.result_flag & Result.WA:
-                        # If we failed a 0-point case, we will short-circuit every case after this.
-                        is_short_circuiting_enabled |= not case.points
+                # If the submission was killed due to a user-initiated abort, any result is meaningless.
+                if self._abort_requested:
+                    raise GradingAbort
 
-                        # Short-circuit if we just failed a case in a batch, or if short-circuiting is currently enabled
-                        # for all test cases (either this was requested by the site, or we failed a 0-point case in the
-                        # past).
-                        is_short_circuiting |= batch_number is not None or is_short_circuiting_enabled
+                if result.result_flag & result.WA:
+                    skip_current = True
 
                 # Legacy hack: we need to allow graders to read and write `proc_output` on the `Result` object, but the
                 # judge controller only cares about the trimmed output, and shouldn't waste memory buffering the full
                 # output. So, we trim it here so we don't run out of memory in the controller.
                 result.proc_output = result.output
-                yield IPC.RESULT, (batch_number, case_number, result)
+                return result
 
-            if batch_number:
-                if not is_short_circuiting:
-                    passed_batches.add(batch_number)
+        def should_run_test(case: TopLevelCase, failed_so_far: Set[int]) -> bool:
+            if case.dependencies is None:
+                # Default: depends on nothing.
+                return True
+            else:
+                return len(case.dependencies & failed_so_far) == 0
+
+        failed_pretests: Set[int] = set()
 
-                yield IPC.BATCH_END, (batch_number,)
-                is_short_circuiting &= is_short_circuiting_enabled
+        def should_run_main_test(case: TopLevelCase, failed_so_far: Set[int]) -> bool:
+            good_pretests = False
```

Comment:
Don't need this variable here, it's defined in both branches.

Reply:
Fixed